jib-cli/src/test/java/com/google/cloud/tools/jib/cli/cli2/CredentialsTest.java

```diff
@@ -0,0 +1,265 @@
+                "--from-credential-helper=abc", "--to-username=ignored", "--to-password=ignored"
+              }
+            },
+          });
+    }
+
+    @Parameterized.Parameter public String[] args;
+
+    @Test
+    public void testGetFromCredentialRetriever() throws FileNotFoundException {
+      JibCli buildOptions =
+          CommandLine.populateCommand(new JibCli(), ArrayUtils.addAll(DEFAULT_ARGS, args));
+      Credentials.getFromCredentialRetrievers(buildOptions, defaultCredentialRetrievers);
+      Mockito.verify(defaultCredentialRetrievers).setCredentialHelper("abc");
+      Mockito.verify(defaultCredentialRetrievers).asList();
+      Mockito.verifyNoMoreInteractions(defaultCredentialRetrievers);
+    }
+  }
+
+  @RunWith(Parameterized.class)
+  public static class GetToCredentialRetrieverUsernamePassword {
+    @Rule public MockitoRule mockitoJUnit = MockitoJUnit.rule();
+    @Mock private DefaultCredentialRetrievers defaultCredentialRetrievers;
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+      return Arrays.asList(
+          new Object[][] {
+            {"--username/--password", new String[] {"--username=abc", "--password=xyz"}},
+            {
+              "--to-username/--to-password", new String[] {"--to-username=abc", "--to-password=xyz"}
+            },
+            {
+              "--to-username/--to-password",
+              new String[] {
+                "--to-username=abc",
+                "--to-password=xyz",
+                "--from-username=ignored",
+                "--from-password=ignored"
+              }
+            },
+            {
+              "--to-username/--to-password",
+              new String[] {
+                "--to-username=abc", "--to-password=xyz", "--from-credential-helper=ignored"
+              }
+            },
+          });
+    }
+
+    @Parameterized.Parameter(0)
+    public String expectedSource;
+
+    @Parameterized.Parameter(1)
+    public String[] args;
+
+    @Test
+    public void testGetToCredentialRetriever() throws FileNotFoundException {
+      JibCli buildOptions =
+          CommandLine.populateCommand(new JibCli(), ArrayUtils.addAll(DEFAULT_ARGS, args));
+      Credentials.getToCredentialRetrievers(buildOptions, defaultCredentialRetrievers);
+      ArgumentCaptor<Credential> captor = ArgumentCaptor.forClass(Credential.class);
+      Mockito.verify(defaultCredentialRetrievers)
+          .setKnownCredential(captor.capture(), ArgumentMatchers.eq(expectedSource));
+      assertThat(captor.getValue()).isEqualTo(Credential.from("abc", "xyz"));
+      Mockito.verify(defaultCredentialRetrievers).asList();
+      Mockito.verifyNoMoreInteractions(defaultCredentialRetrievers);
+    }
+  }
+
+  @RunWith(Parameterized.class)
+  public static class GetFromCredentialRetrieverUsernamePassword {
+    @Rule public MockitoRule mockitoJUnit = MockitoJUnit.rule();
+    @Mock private DefaultCredentialRetrievers defaultCredentialRetrievers;
```

Comment:
All these can be `final`.

Reply:
yeah I keep making this mistake, I should be more careful, maybe there's a rule we can add to something